Add NUT device actions

pestevez opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->



## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds device actions to the [NUT integration](https://www.home-assistant.io/integrations/nut/) to run [commands](https://networkupstools.org/docs/user-manual.chunked/apcs03.html) on the UPS.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR is related to this feature request: [Nut Commands (upscmd)](https://community.home-assistant.io/t/nut-commands-upscmd/401107).
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/24698
- I tweaked the autogenerated tests.
- Only parameterless commands are supported because the dependent library does not support sending parameters to the command.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mdegat01 commented on homeassistant/components/nut/services.yaml:
```diff
@@ -0,0 +1,18 @@
+run_command:
+  # Service name as shown in UI
+  name: Run command
+  # Description of the service
+  description: Invoke a given instant command in the UPS unit
+  fields:
+    # Key of the field
+    name:
+      # Field name as shown in UI
+      name: Name
+      # Description of the field
+      description: Command name
+      # Whether or not field is required (default = false)
+      required: true
+      # Example value that can be passed for this field
+      example: "beeper.disable"
+      # The default field value
+      default: ""
```
Is it possible to map the commands to individual services? Like rather then having a single `run_command` service that requires users go look in the NUT documentation to see what values are allowed for `name`, could we have `nut.beeper_disable`, `nut.beeper_enable`, `nut.beeper_on`, etc. services?

Also what about other inputs? This service needs to accept a `target` of a NUT type device for one. The user may have HA connected to multiple UPS's. If so it's impossible tell which one to send the command to right now.

But besides that, do any of the commands accept other inputs? If so then I would strongly consider taking the approach I suggested above of mapping the commands to individual services rather then having one generic service. This will get very confusing if you start adding additional fields like `input_1`, `input_2`, etc. to handle the possibility that a command might have additional inputs.

mdegat01 replied:
Ideally these should be device actions as well since the NUT integration creates a device for each connected UPS.

pestevez replied:
Thanks for the quick response, @mdegat01. Yes, your proposed approach of registering a service for each command is definitely a way that we can go. The main challenge is that the supported commands vary by UPS model, meaning that the list of all supported services is dynamic (although I assume that it is always a subset of those described [here](https://networkupstools.org/docs/user-manual.chunked/apcs03.html)).

What I had in mind was exposing later on a sensor with the list of all the supported commands by each unit, so that it was clear which were the valid values for the command name.

Regarding your proposal, I see a few options, but please let me know if some of them don't make sense as I'm still familiarizing myself with HA:

1. Have a list of common commands and register those as all potential services.
   1. We could also keep the generic `run_command` service as a fallback option for those commands not on the list.
3. Dynamically register services based on the available commands for the units.
   1. That limits our ability to create the `services.yaml` file and I don't know how we would handle having multiple devices with commands that are available to only one of them.
5. Keeping just the existing `run_command` service.

Some commands do accept a single parameter, according to [the documentation](https://networkupstools.org/docs/developer-guide.chunked/ar01s09.html#:~:text=%3Ccmdparam%3E%20is%20an%20additional%20and%20optional%20parameter%20for%20the%20command.), although it is unclear to me which of them are. If we go with the option of the generic service, then we can have a generic optional parameter `cmdparam` or similar. I was planning on tackling this a future enhancement.

What do you think? Which of the three options should we go with?

I will fix the missing target, good catch. I need to test it using a system with multiple UPS devices to ensure that it looks good.

I will also take a look at the device actions. It was not very clear to me from the docs how to use them in this integration, but I will take another pass at it.

mdegat01 replied:
Is there a way to find out what commands are supported by a particular model by making an API call? If so then that's a strong argument for device actions IMO.

I believe you'll have to register services for all the possible commands to do this and then validate whether the target supports the command based on device metadata from that API call. But what you can do is register different device actions when creating the device. So only the commands that make sense for that model show up in the device action UI making mistakes impossible there.

As for whether to have a fallback generic `run_command` service, I guess it depends how many commands there are. If there's like 20 or so then probably just plan to make the common ones into services now and more over time on user request. If there's hundreds and the list grows regularly then I would agree with a fallback generic `run_command` service as it's unlikely we'll keep up with that. Should also let the code owner weigh in here and they'll be supporting this.

pestevez replied:
> Is there a way to find out what commands are supported by a particular model by making an API call? 

Yes. Device actions would be the best way to go then per your recommendation.

> f there's like 20 [commands] or so then probably just plan to make the common ones into services now and more over time on user request.

That's the case.

> Should also let the code owner weigh in here

@bdraco, @ollo69 - any thoughts?

bdraco replied:
It looks like non of these commands take arguments?

If that is the case they might be a bit more user friendly if you implement them as button entities

bdraco replied:
I guess some of them do take params

```
INSTCMD <upsname> <cmdname> [<cmdparam>]
INSTCMD su700 test.panel.start
INSTCMD su700 load.off.delay 120
```

bdraco replied:
I think we end up with a whole lot of buttons or services and that's not particularly a great experience.

I think device actions is the way to go here.

jonferreira commented:
pretty happy to see this going forward, just waiting to be published so I can test it

bdraco reviewed (commented):
Please see comment above

bdraco commented on homeassistant/components/nut/device_action.py:
```diff
@@ -0,0 +1,113 @@
+from . import PyNUTData
+from .const import DEVICE_SUPPORTED_COMMANDS, DOMAIN, PYNUT_DATA
+
+ACTION_TYPES = {
```
Can we generate this from INTEGRATION_SUPPORTED_COMMANDS by mutating that data?

pestevez replied:
That was my original intent, since it feels duplicated. After a few tries I gave up because I thought that `ACTION_TYPES` was expected to be a constant and that I could not generate the constant using a mutation. I am fairly new to Python and totally new to Home Assistant development, so I could be totally wrong here.

As a side note, with `INTEGRATION_SUPPORTED_COMMANDS` I was also aiming at filtering out those that receive parameters, since the library used does not support params. I could try to submit a PR later on if we feel it makes sense to support those two missing commands, but I am unclear of whether the library is still maintained.

By the way, I am still working on a few tests and may revisit some of the code, but your feedback is greatly appreciated.

bdraco replied:
Untested but I think you could write it as 
```diff
diff --git a/homeassistant/components/nut/device_action.py b/homeassistant/components/nut/device_action.py
index 1cd6d5275c..a288fbd781 100644
--- a/homeassistant/components/nut/device_action.py
+++ b/homeassistant/components/nut/device_action.py
@@ -10,40 +10,15 @@ import homeassistant.helpers.config_validation as cv
 from homeassistant.helpers.typing import ConfigType, TemplateVarsType
 
 from . import PyNUTData
-from .const import DEVICE_SUPPORTED_COMMANDS, DOMAIN, PYNUT_DATA
-
-ACTION_TYPES = {
-    "beeper_disable",
-    "beeper_enable",
-    "beeper_mute",
-    "beeper_toggle",
-    "bypass_start",
-    "bypass_stop",
-    "calibrate_start",
-    "calibrate_stop",
-    "load_off",
-    "load_on",
-    "outlet_n_load_cycle",
-    "outlet_n_load_off",
-    "outlet_n_load_on",
-    "outlet_n_shutdown_return",
-    "reset_input_minmax",
-    "reset_watchdog",
-    "shutdown_reboot",
-    "shutdown_reboot_graceful",
-    "shutdown_return",
-    "shutdown_stayoff",
-    "shutdown_stop",
-    "test_battery_start",
-    "test_battery_start_deep",
-    "test_battery_start_quick",
-    "test_battery_stop",
-    "test_failure_start",
-    "test_failure_stop",
-    "test_panel_start",
-    "test_panel_stop",
-    "test_system_start",
-}
+from .const import (
+    DEVICE_SUPPORTED_COMMANDS,
+    DOMAIN,
+    PYNUT_DATA,
+    INTEGRATION_SUPPORTED_COMMANDS,
+)
+
+ACTION_TYPES = {cmd.replace(".", "_") for cmd in INTEGRATION_SUPPORTED_COMMANDS}
+
 
 ACTION_SCHEMA = cv.DEVICE_ACTION_BASE_SCHEMA.extend(
     {

```

pestevez replied:
Apologies for the radio silence, @bdraco. Your proposal works like a charm. The PR should be good for another pass. Thanks!

bdraco commented:
I think we are just missing test coverage and maybe some minor cleanups at this point

jonferreira commented:
almost there?

pestevez committed:
Add NUT device actions from commands

jonferreira commented:
does this means this is now available on 2023.1.6?

pestevez commented:
No, @jonferreira, not yet. I am missing a couple of spots in the test coverage but I'll need to make some refactoring for those. Once that is done, and assuming there is no additional feedback to address, we'll merge this branch into `dev` so that it can be available in the next release.

jonferreira commented:
ah right, got it @pestevez 
with luck it will be on 2023.2 release then :)

Bluscream commented:
I'd love to see this available in the future (being able to turn on/off the beeper and load)

pestevez commented:
@bdraco - I got the test coverage to 100% and made some edits, so it should be good to go. Let me know if you want me tackle any specific edits. Thanks!

pestevez commented:
@bdraco - My apologies, I realized that I squashed all the commits, so it is harder to review the new changes. Let me know if you want me to break them up, since I don't think it'd be too difficult.

I basically refactored the test cases to support overriding the `list_commands` function.

bdraco commented on homeassistant/components/nut/__init__.py:
```diff
@@ -79,6 +81,22 @@ async def async_update_data() -> dict[str, str]:
+    all_device_supported_commands = data.list_commands() or {}
+    supported_commands: set[str] = set()
+
+    for [
```
The [] here can be dropped

pestevez replied:
I removed the `[]` but after saving the file but then`()` are automatically added; I left the latter. Please let me know if there is a better way to do it. Thanks!

bdraco replied:
() is faster than [] so all fine there

bdraco commented on homeassistant/components/nut/device_action.py:
```diff
@@ -0,0 +1,84 @@
+    """List device actions for Network UPS Tools (NUT) devices."""
+    actions = []
+    entry_id = _get_entry_id_from_device_id(hass, device_id)
+    if entry_id is not None:
```
Reverse the conditional and return right away if it's not met

bdraco commented on homeassistant/components/nut/device_action.py:
```diff
@@ -0,0 +1,84 @@
+    command_name = _get_command_name(device_action_name)
+    device_id: str = config[CONF_DEVICE_ID]
+
+    hass.async_add_executor_job(_run_command, hass, device_id, command_name)
```
Missing await

bdraco commented on homeassistant/components/nut/__init__.py:
```diff
@@ -270,3 +289,18 @@ def update(self) -> None:
+    def list_commands(self) -> dict[str, str] | None:
+        """Fetch the list of supported commands."""
+        try:
+            return dict[str, str](self._client.list_commands(self._alias))
```
We should probably cast here instead

bdraco commented on homeassistant/components/nut/__init__.py:
```diff
@@ -270,3 +289,18 @@ def update(self) -> None:
+        try:
+            self._client.run_command(self._alias, command_name)
+        except PyNUTError as err:
+            _LOGGER.error("Error running command %s, %s", command_name, err)
```
This should probably raise HomeAssistsntError instead.

bdraco commented on homeassistant/components/nut/__init__.py:
```diff
@@ -270,3 +289,18 @@ def update(self) -> None:
         if self._device_info is None:
             self._device_info = self._get_device_info()
+
+    def run_command(self, command_name: str | None) -> None:
```
It would be nicer to make this async_run_command and do the executor call here

pestevez committed:
Address feedback

pestevez commented:
I will rebase to remove the merge conflicts.

bdraco committed:
Merge branch 'dev' into nut-run-command

bdraco commented on homeassistant/components/nut/device_action.py:
```diff
@@ -0,0 +1,81 @@
+    for command_name in supported_commands:
+        device_action_name = _get_device_action_name(command_name)
+
+        base_action = {
+            CONF_DEVICE_ID: device_id,
+            CONF_DOMAIN: DOMAIN,
+        }
```
This never changes so it should move outside of the loop above

pestevez replied:
Great catch! I had copied it from somewhere else but I had not realized that after I tweaked it to not include the entry then it could be moved outside.

bdraco commented on tests/components/nut/test_device_action.py:
```diff
@@ -0,0 +1,222 @@
+    return mock_device_registry(hass)
+
+
+@pytest.fixture
```
Drop this and we can use the built in `entity_registry` fixture instead

bdraco committed:
Merge branch 'dev' into nut-run-command

pestevez committed:
Move base_action outside of loop

pestevez committed:
Clean up unnecessary registry mocks in nut test

pestevez commented:
All yours @bdraco. I modified the test to match #87732 as much as possible.

Thanks for all the hand holding, hopefully the next PR will be easier to review!

bdraco commented on tests/components/nut/test_device_action.py:
```diff
@@ -0,0 +1,201 @@
+        },
+    )
+
+    with mock.patch("logging.Logger.error") as error:
```
We shouldn't patch the logger.  Instead if you want to assert something is logged, use the caplog fixture

pestevez replied:
I still have pending fixing this one. I'll get back to it shortly.

bdraco commented on homeassistant/components/nut/device_action.py:
```diff
@@ -0,0 +1,81 @@
+
+def _get_entry_id_from_device_id(hass: HomeAssistant, device_id: str) -> str | None:
+    registry = device_registry.async_get(hass)
+    device = registry.async_get(device_id)
```
You could use a walrus here

bdraco commented on homeassistant/components/nut/device_action.py:
```diff
@@ -0,0 +1,81 @@
+        CONF_DEVICE_ID: device_id,
+        CONF_DOMAIN: DOMAIN,
+    }
+    for command_name in supported_commands:
```
actions could be an anonymous list comprehension instead

bdraco commented on homeassistant/components/nut/device_action.py:
```diff
@@ -0,0 +1,81 @@
+    hass: HomeAssistant, device_id: str
+) -> list[dict[str, str]]:
+    """List device actions for Network UPS Tools (NUT) devices."""
+    entry_id = _get_entry_id_from_device_id(hass, device_id)
```
Could use a walrus here but it might be more lines

bdraco commented on homeassistant/components/nut/__init__.py:
```diff
@@ -79,6 +83,22 @@ async def async_update_data() -> dict[str, str]:
     status = coordinator.data
 
+    all_device_supported_commands = data.list_commands() or {}
+    supported_commands: set[str] = set()
```
I would probably write this as a comprehension since I don't think there is much value in the debug logging as long as it's covered by a test

pestevez replied:
The intent was to help to troubleshoot in case that some expected commands were not shown in the UI as available options, but we can add it back later on if it ends up being an issue.

bdraco reviewed (commented):
There are a few opportunities to make this a bit more succinct. I'll put this on my production and do some testing after the next turn.

Thanks!

pestevez committed:
Use walrus

pestevez committed:
Use comprehensions

bdraco commented on homeassistant/components/nut/device_action.py:
```diff
@@ -30,23 +30,20 @@ async def async_get_actions(
-        actions.append({**base_action, CONF_TYPE: device_action_name})
-
-    return actions
+    return (
```
This one didn't turn out how I was envisioning.  I'll put up a suggestion shortly 👍

pestevez replied:
You were faster than me. I meant to reply saying that I could have gone too far on making the code more succinct, at the expense of code readability. Your suggestion fixes that. Thanks!

bdraco commented on homeassistant/components/nut/device_action.py:
```diff
@@ -0,0 +1,76 @@
+    hass: HomeAssistant, device_id: str
+) -> list[dict[str, str]]:
+    """List device actions for Network UPS Tools (NUT) devices."""
+    return (
+        []
+        if (entry_id := _get_entry_id_from_device_id(hass, device_id)) is None
+        else [
+            {
+                **{
+                    CONF_DEVICE_ID: device_id,
+                    CONF_DOMAIN: DOMAIN,
+                },
+                CONF_TYPE: _get_device_action_name(command_name),
+            }
+            for command_name in hass.data[DOMAIN][entry_id][DEVICE_SUPPORTED_COMMANDS]
+        ]
+    )
```
```suggestion
    if (entry_id := _get_entry_id_from_device_id(hass, device_id)) is None:
        return []
    base_action = {
        CONF_DEVICE_ID: device_id,
        CONF_DOMAIN: DOMAIN,
    }
    supported_commands = hass.data[DOMAIN][entry_id][DEVICE_SUPPORTED_COMMANDS]
    return [
        {CONF_TYPE: _get_device_action_name(command_name)} | base_action
        for command_name in supported_commands
    ]
```

bdraco commented:
Thank you for your contribution thus far! 🎖 Since this is a significant contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️ 

Please, add your GitHub username to the `manifest.json` of this integration.

For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

bdraco commented on homeassistant/components/nut/const.py:
```diff
@@ -38,3 +40,67 @@
+COMMAND_CALIBRATE_STOP = "calibrate.stop"
+COMMAND_LOAD_OFF = "load.off"
+COMMAND_LOAD_ON = "load.on"
+COMMAND_OUTLET_N_LOAD_CYCLE = "outlet.n.load.cycle"
+COMMAND_OUTLET_N_LOAD_OFF = "outlet.n.load.off"
+COMMAND_OUTLET_N_LOAD_ON = "outlet.n.load.on"
+COMMAND_OUTLET_N_SHUTDOWN_RETURN = "outlet.n.shutdown.return"
```
I think these take a specific outlet as the N value?

pestevez replied:
You are right, my bad. These were not supported by my unit and I missed the parameter from the docs.

pestevez committed:
Improve readability of async_get_actions

Co-authored-by: J. Nick Koston <nick@koston.org>

pestevez committed:
Remove unsupported outlet specific commands

pestevez committed:
Add pestevez to list of NUT code owners

pestevez commented:
> Thank you for your contribution thus far! 🎖 Since this is a significant contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️
> 
> Please, add your GitHub username to the `manifest.json` of this integration.
> 
> For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

Thanks for the opportunity to support this project!

pestevez committed:
Use caplog

pestevez committed:
Merge branch 'dev' into nut-run-command

pestevez committed:
Merge branch 'dev' into nut-run-command

bdraco committed:
Merge branch 'dev' into nut-run-command

bdraco committed:
update for new requirements

bdraco commented:
I updated the PR for the latest requirements

bdraco commented:
I created a script to run the `beeper.mute` command and received and error while testing it

```
2023-03-07 04:31:37.177 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [140579372022320] Error running command beeper.mute, ERR USERNAME-REQUIRED
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/nut/__init__.py", line 290, in async_run_command
    await hass.async_add_executor_job(
  File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.10/site-packages/pynut2/nut2.py", line 503, in run_command
    raise PyNUTError(result.replace("\n", ""))
pynut2.nut2.PyNUTError: ERR USERNAME-REQUIRED

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 199, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1822, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1859, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/usr/src/homeassistant/homeassistant/components/script/__init__.py", line 474, in _service_handler
    await self.async_turn_on(variables=service.data, context=service.context)
  File "/usr/src/homeassistant/homeassistant/components/script/__init__.py", line 433, in async_turn_on
    await coro
  File "/usr/src/homeassistant/homeassistant/components/script/__init__.py", line 463, in _async_run
    return await self.script.async_run(script_vars, context)
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 1524, in async_run
    await asyncio.shield(run.async_run())
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 409, in async_run
    await self._async_step(log_exceptions=False)
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 453, in _async_step
    self._handle_exception(
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 476, in _handle_exception
    raise exception
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 451, in _async_step
    await getattr(self, handler)()
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 692, in _async_device_step
    await device_action.async_call_action_from_config(
  File "/usr/src/homeassistant/homeassistant/components/device_automation/action.py", line 71, in async_call_action_from_config
    await platform.async_call_action_from_config(hass, config, variables, context)
  File "/usr/src/homeassistant/homeassistant/components/nut/device_action.py", line 60, in async_call_action_from_config
    await data.async_run_command(hass, data, command_name)
  File "/usr/src/homeassistant/homeassistant/components/nut/__init__.py", line 294, in async_run_command
    raise HomeAssistantError(
homeassistant.exceptions.HomeAssistantError: Error running command beeper.mute, ERR USERNAME-REQUIRED
```

pestevez commented:
Can you try to run the action from the UI of a new automation instead to see if the error persists? I am seeing in the logs that username and password are being fetched from the config and sent correctly to the nut server.

I am trying to determine whether the issue is related to the way in which the device action is executed.

<img width="595" alt="image" src="https://user-images.githubusercontent.com/18663703/223778769-376843bf-6027-43e7-bc2c-3f690dc0d4a1.png">

<img width="692" alt="image" src="https://user-images.githubusercontent.com/18663703/223779136-a7294f5c-cbe9-4abc-94ab-21494f655a2a.png">

pestevez commented:
@bdraco - I ran some additional tests and was unable to reproduce the issue. It calls my attention that the integration is complaining about the username parameter not being present, since the logic to send the credentials has not changed.

A few questions that will help me to debug?

1. Do you have a username and password configured in the client for the device that you are testing or do you have them blank?
2. Are you receiving updates for the device sensors e.g. `Load` and `Status`?
3. If you are using the [NUT add-on](https://github.com/hassio-addons/addon-nut), do you have `instcmds` set to `all`?

pestevez committed:
Remove unused parameter

pestevez committed:
Remove strings for unsupported commands

bdraco committed:
Merge branch 'dev' into nut-run-command

bdraco committed:
Merge branch 'dev' into nut-run-command

bdraco committed:
Merge branch 'dev' into nut-run-command

bdraco commented:
> @bdraco - is there anything else required from my end here? If so, would you mind answering the questions above so that I can troubleshoot? Thanks

Below is the config entry.  I'm running it on core on a synology dsm box v6 (not HAOS) and upsd is running on the host.

```
      {                    
        "entry_id": "6038c7ab47214a3a9e6951a9052f9833",
        "version": 1,
        "domain": "nut",
        "title": "localhost:3493",
        "data": {               
          "host": "localhost",  
          "port": 3493
        },                      
        "options": {
          "scan_interval": 60
        },                        
        "pref_disable_new_entities": false,
        "pref_disable_polling": false, 
        "source": "user",             
        "unique_id": null,
        "disabled_by": null   
      }, 
```

pestevez committed:
Ensure user credentials to enable commands

pestevez commented:
@bdraco - please let me know what you think about this approach.

bdraco commented:
started the CI. Will try to find some time to do another review this weekend

pestevez committed:
Run prettier on strings file

pestevez commented:
Cool, no rush. Take your time. I updated the documentation to explain this approach.

I also fixed the linter that was complaining about the strings file.

bdraco commented:
I'm traveling this week but will try to take a look when I can

https://developers.home-assistant.io/docs/integration_quality_scale_index?_highlight=silver#silver-

> Operations like service calls and entity methods (e.g. Set HVAC Mode) have proper exception handling. Raise ValueError on invalid user input and raise HomeAssistantError for other failures such as a problem communicating with a device.

bdraco committed:
Merge branch 'dev' into nut-run-command

bdraco committed:
unpack

bdraco commented:
Thanks @pestevez
